### PR TITLE
RFC: commitlog allocation/deletion/flush request rate counters + footprint projection

### PR DESCRIPTION
**Pull request by elcallio**
Adds measuring the apparent delta vector of footprint added/removed within
the timer time slice, and potentially include this (if influx is greater
than data removed) in threshold calculation. The idea is to anticipate
crossing usage threshold within a time slice, so request a flush slightly
earlier, hoping this will give all involved more time to do their disk
work.

Obviously, this is very akin to just adjusting the threshold downwards,
but the slight difference is that we take actual transaction rate vs.
segment free rate into account, not just static footprint.

Note: this is a very simplistic version of this anticipation scheme,
we just use the "raw" delta for the timer slice.
A more sophisiticated approach would perhaps do either a lowpass
filtered rate (adjust over longer time), or a regression or whatnot.
But again, the default persiod of 10s is something of an eternity,
so maybe that is superfluous...

### CODE REVIEW

File: db/commitlog/commitlog.cc [hunk 1]
```diff
@@ -1512,28 +1512,42 @@ void db::commitlog::segment_manager::flush_segments(uint64_t size_to_remove) {
...
-            if (n <= s->size_on_disk()) {
-                high = replay_position(s->_desc.id, db::position_type(s->size_on_disk()));
-                break;
-            }
-            n -= s->size_on_disk();
+    uint64_t n = size_to_remove;
+    uint64_t flushing = 0;
+
+    for (auto& s : _segments) {
+        if (s->is_still_allocating()) {
```
Comment (bhalevy) [inline db/commitlog/commitlog.cc:1569]: This seems to be a behavioral change and it is unclear why we're breaking from the loop here.

**Review by bhalevy — COMMENTED** [review]
Also typo in "commitlog: Add (internal) measurement of byte rates add/release/flush"
s/persiod/period/

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:1569]: It is not really, just applying the same logic as (previously) above w.r. high rp. Still allocating segments cannot be released, i.e. is pointless to include in a flush request. The allocating segment will also always be the last one, so in effect we do almost always end at end()-1, unless things are shut/shutting down, in which case we could have _no_ allocating segment.

Comment (bhalevy) [reply db/commitlog/commitlog.cc:1569]: @elcallio so let's add some comments explaining those decisions in the code.
As someone that isn't intimately familiar with the implementation I couldn't find my way around...

Comment (bhalevy) [reply db/commitlog/commitlog.cc:1569]: Otherwise LGTM

Comment (elcallio) [reply db/commitlog/commitlog.cc:1573]: Fixed.

File: db/commitlog/commitlog.cc [hunk 2]
```diff
@@ -1513,28 +1513,45 @@ void db::commitlog::segment_manager::flush_segments(uint64_t size_to_remove) {
...
-            if (n <= s->size_on_disk()) {
-                high = replay_position(s->_desc.id, db::position_type(s->size_on_disk()));
-                break;
-            }
-            n -= s->size_on_disk();
+    uint64_t n = size_to_remove;
+    uint64_t flushing = 0;
+
+    for (auto& s : _segments) {
+        // if a segment is allocating, it should be included in flush request,
```
Comment (bhalevy) [inline db/commitlog/commitlog.cc:1570]: s/should/should not/ maybe?

**Review by bhalevy — COMMENTED** [review]
one nit in comment
otherwise looks good

**Comment by elcallio** [comment]
ping?

File: db/commitlog/commitlog.cc [hunk 3]
```diff
@@ -325,13 +325,47 @@ class db::commitlog::segment_manager : public ::enable_shared_from_this<segment_
     requires std::derived_from<T, db::commitlog::entry_writer> && std::same_as<R, decltype(std::declval<T>().result())>
     future<R> allocate_when_possible(T writer, db::timeout_clock::time_point timeout);
 
-    struct stats {
+    template<typename T>
+    struct byte_flow {
+        T bytes_written = 0;
+        T bytes_released = 0;
+        T bytes_flush_requested = 0;
```
Comment (avikivity) [inline db/commitlog/commitlog.cc:332]: Why a template? Isn't uint64_t sufficient?

File: db/commitlog/commitlog.cc [hunk 4]
```diff
@@ -347,6 +381,10 @@ class db::commitlog::segment_manager : public ::enable_shared_from_this<segment_
     };
 
     stats totals;
+    byte_flow<uint64_t> last_bytes;
+    byte_flow<double> bytes_rate;
```
[the remark this answers is not in the document]
Comment (elcallio) [reply db/commitlog/commitlog.cc:385]: To avoid truncation for calculating rates.

**Pull request closed without merging**